### ISSUE TITLE
Private final variables during caching

### ISSUE THREAD

**Issue by brennanleez-coder** [nus-cs2030/2021-s2#189]
Hello guys, I have a question regarding whether my variables should be made private final or not. For the context of Lazy.java, it is not possible to make my cache private final as I would need to store my computed value within it so it can be called later on without computation again. Also, my attribute T t also cannot be private final as my get() method would need to assign the cache to it if it is present and if not null.
can anyone provide some clarity as to when would it be allowed.
I understand that codecrunch flags out attributes that are not private final.

**Comment by LongJiAn99** [comment]
Hi @brennanleez-coder ! Great question, it may be confusing and I am not sure if this is indeed the right answer but as what Prof. Henry mentioned during lecture, I am sure that cache is one of the only exceptions where we do not have to set the attribute as final. In all other cases, I would assume that the attribute has to be final for the sake of immutability. Should there be a need to use cache in the examinations or assignments I would be confident that the question would highlight this and allow us to not set it as a final attribute. Happy coding!

**Comment by ryyant** [comment]
Indeed, what we have been practicing is that all attributes should be set as private final. However, i agree with @LongJiAn99 that there are times where there are bound to be exceptions. I feel that this would be a case where it doesnt have to be final! Hope im right. Good luck!

**Comment by nicolaineo** [comment]
Thanks for bringing this question up! I also have encountered the same problem multiple times where I felt things would be easier is the field did not have to be private final. However I think prof is trying to instill good coding practices into us and thus try to adhere to private final fields.

**Comment by brendancjz** [comment]
Hi Brennan, great question. Yes, in the lecture notes, Prof did not make Cache private final to illustrate the cool feature Optional has with the method .orElseGet(Supplier<? extends T> sup). However, as we are trying to adhere to using private final always, we cannot replicate what Prof did in his example slides. What we can do instead is to have a method within the same class that first checks with the cache is empty, aka Optional.empty. If it is empty, then returns a new object class with the updated cache value from sup.get(). This means that you need to have a constructor that takes in this value from sup.get() and update your cache to be this value. Hope this helps.
